lib/HTFeed/StorageZipAudit.pm

```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/perl
+package HTFeed::StorageZipAudit;
+
+use strict;
+use warnings;
+use Carp;
+use File::Temp;
+use FindBin;
+use Log::Log4perl qw(get_logger);
+
+use HTFeed::Config qw(get_config);
+use HTFeed::DBTools qw(get_dbh);
+use HTFeed::Volume;
+
+sub new {
+  my $class = shift;
+
+  my $self = {
+    @_
+  };
+
+  die("Missing required argument 'storage_name'")
+    unless $self->{storage_name};
+
+  bless($self, $class);
+  return $self;
+}
+
+# Choose and if necessary request a random object for this storage class.
+sub random_object {
+  my $self = shift;
+
+  my $obj;
+  my $sql = 'SELECT namespace,id,path,version,saved_md5sum FROM feed_backups' .
+            ' WHERE storage_name=?' .
+            ' AND deleted IS NULL' .
+            ' AND saved_md5sum IS NOT NULL' .
+            ' AND restore_request IS NULL' .
+            ' ORDER BY RAND() LIMIT 1';
+  eval {
+    my $ref = get_dbh->selectall_arrayref($sql, undef, $self->{storage_name});
+    if (scalar @$ref) {
+      $obj = $self->storage_object(@{$ref->[0]});
+    }
+  };
+  if ($@) {
+    die "Database query failed: $@";
+  }
+  return $obj;
+}
+
+
+# Get the object(s) selected for audit by choose_random_object() or some other criteria.
+# In the case of AWS Glacier, it is one or more objects selected in a previous run
+# that have been restored in the intervening time.
+# These are the objects on the filesystem ready for checksumming.
+sub all_objects {
+  my $self = shift;
+
+  return [$self->random_object()];
+}
+
+sub storage_object {
+  my $self      = shift;
+  my $namespace = shift;
+  my $objid     = shift;
+  my $path      = shift;
+  my $version   = shift;
+  my $md5       = shift;
+  
+  unless ($namespace && $objid && $path && $version && $md5) {
+    croak "invalid storage_object() args: namespace, objid, path, version, md5 required";
+  }
+  my $obj = {namespace => $namespace, objid => $objid, version => $version,
+             path => $path, storage_name => $self->{storage_name},
+             md5 => $md5};
+  $obj->{volume} = HTFeed::Volume->new(packagetype => 'pkgtype', namespace => $namespace, objid => $objid);
+  my $config = get_config('storage_classes');
+  $obj->{storage_config} = $config->{$self->{storage_name}};
+  $obj->{storage} = $obj->{storage_config}->{class}->new(volume    => $obj->{volume},
```

Comment:
Blargh. This ends up more complicated than I'd like, since storage is specific to a particular volume, but the zip audit isn't. I think there is a missing concept here of the overall storage instance as opposed to the storage for a particular volume (or rather, that concept only exists via name rather than as a particular class instance.) I don't think that's worth addressing now, but I made a note on HT-2521.

Reply:
Another option could be splitting up `ZipAuditor` into two pieces. One piece would take the storage name, pick a volume using that storage, and fetch info from the database about the volume. That's independent of the particular storage backend. The second piece, which could be bound to a volume and returned by the storage backend, would do the actual verification work. That might be easier to do now without big changes to `Storage`.

Reply:
I've gone ahead and worked a bit on merging the zip audits with the filesystem crawls. I'm implementing iterators, for example, as Storage class methods rather than going with the route of needing a storage instance bound to a volume. Until we're actually talking about volumes, that is. It's maybe too early to tell if that's going to work well, but I may be able to scrape away some of the complexity.

In the case you cite, I've moved storage_config wrangling rom the storage_object to the StorageZipAudit initializer.

Some of the path parsing functionality from audit/DataDenPath.pm looks like a good fit for PrefixedVersions.pm as one or more class methods.